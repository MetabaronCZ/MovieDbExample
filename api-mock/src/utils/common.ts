import { Request } from 'express';

export type QueryParam = Request['query'][number];
export type Query = Record<string, QueryParam>;

const wait = (delay: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const requestDelay = (cb: () => void): void => {
  const delay = 800 + Math.floor(400 * Math.random());
  wait(delay)
    .then(cb)
    .catch((error: unknown) => {
      console.error(error);
    });
};

// extract simple string value from Express query parameter result
export const parseQueryParam = (param: QueryParam): string => {
  const value = Array.isArray(param) ? param[0] : param;
  return 'string' === typeof value ? value : '';
};

type Entries<T extends Record<string, unknown>> = {
  readonly [K in keyof T]: [K, T[K]];
}[keyof T][];

// extract object entries, correctly typed
export const getObjectEntries = <T extends Record<string, unknown>>(
  obj: T,
): Entries<T> => {
  return Object.entries(obj) as Entries<T>;
};
