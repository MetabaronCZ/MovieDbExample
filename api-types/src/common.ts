export const perPages = [25, 50, 100] as const;
export const defaultPerPage = perPages[0];

export const sortDirections = ['ascending', 'descending'] as const;
export type SortDirection = (typeof sortDirections)[number];
