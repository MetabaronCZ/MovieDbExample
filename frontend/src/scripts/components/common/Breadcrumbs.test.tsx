import { getAllByRole, getByRole } from '@testing-library/dom';
import { describe, expect, it } from '@jest/globals';

import { Breadcrumbs } from 'components/common/Breadcrumbs';
import { TestComponentWrapper, waitForComponent } from 'test-utils/component';

describe('components/common/Breadcrumbs', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <Breadcrumbs items={['movieList', 'movieDetail']} />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const items = getAllByRole(list, 'listitem');
    expect(items.length).toEqual(3);

    const itemContents = items.map((item) => item.textContent);
    expect(itemContents).toEqual(['DomÅ¯', 'Filmy', 'Detail filmu']);
  });

  it('should render empty list', async () => {
    const { container } = await waitForComponent(<Breadcrumbs />, {
      wrapper: TestComponentWrapper,
    });
    expect(container.innerHTML).toEqual('');
  });
});
