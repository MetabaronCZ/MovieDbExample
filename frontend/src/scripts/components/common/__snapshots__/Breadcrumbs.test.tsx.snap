// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/common/Breadcrumbs should render properly 1`] = `
.c3 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
}

.c3:hover {
  text-decoration: none;
}

.c3:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c2 {
  display: flex;
  flex-direction: row;
  gap: 4px;
  align-items: center;
}

.c0 {
  list-style-type: none;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
}

.c1:not(:first-child)::before {
  content: '•';
  padding: 0 8px;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-house fa-sm "
          data-icon="house"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M575.8 255.5c0 18-15 32.1-32 32.1l-32 0 .7 160.2c0 2.7-.2 5.4-.5 8.1l0 16.2c0 22.1-17.9 40-40 40l-16 0c-1.1 0-2.2 0-3.3-.1c-1.4 .1-2.8 .1-4.2 .1L416 512l-24 0c-22.1 0-40-17.9-40-40l0-24 0-64c0-17.7-14.3-32-32-32l-64 0c-17.7 0-32 14.3-32 32l0 64 0 24c0 22.1-17.9 40-40 40l-24 0-31.9 0c-1.5 0-3-.1-4.5-.2c-1.2 .1-2.4 .2-3.6 .2l-16 0c-22.1 0-40-17.9-40-40l0-112c0-.9 0-1.9 .1-2.8l0-69.7-32 0c-18 0-32-14-32-32.1c0-9 3-17 10-24L266.4 8c7-7 15-8 22-8s15 2 21 7L564.8 231.5c8 7 12 15 11 24z"
            fill="currentColor"
          />
        </svg>
        <a
          class="c3"
          data-discover="true"
          href="/"
        >
          Domů
        </a>
      </div>
    </li>
    <li
      class="c1"
    >
      <a
        class="c3"
        data-discover="true"
        href="/movies"
      >
        Filmy
      </a>
    </li>
    <li
      class="c1"
    >
      <strong>
        Detail filmu
      </strong>
    </li>
  </ul>
</div>
`;
